### PR TITLE
feat: improves error messaging for better debugging

### PR DESCRIPTION
**Pull request by dutterbutter**
# What :computer: 
* Adds formatting for printing structured error messaging
* Prints structured error logs when revert or halt is hit
* Related to #436 

# Why :hand:
* Previous error messages were confusing / did not provide insightful messages

# Evidence :camera:

**Contract reversion error:**
Screenshot 2024-12-03 at 12 34 29 PM

**Paymaster error:**
Screenshot 2024-12-03 at 12 35 01 PM

**Function selector error:**
Screenshot 2024-12-03 at 12 35 21 PM

### CODE REVIEW

File: src/node/in_memory.rs [hunk 1]
```diff
@@ -795,7 +786,38 @@ impl<S: std::fmt::Debug + ForkSource> InMemoryNodeInner<S> {
...
+        let tracers = vec![
+            CallErrorTracer::new(error_flags_result.clone()).into_tracer_pointer(),
+            CallTracer::new(call_tracer_result.clone()).into_tracer_pointer(),
+            BootloaderDebugTracer {
+                result: bootloader_debug_result.clone(),
+            }
+            .into_tracer_pointer(),
+        ];
+        // TODO: Ask if vm.inspect is the same as vm.execute
+        let tx_result = vm.inspect(&mut tracers.into(), InspectExecutionMode::OneTx);
```
Comment (dutterbutter) [inline src/node/in_memory.rs:802]: @itegulov do you know if calling `vm.inspect` vs `vm.execute` makes any difference? It doesnt seem so but wanted to confirm.

File: src/node/in_memory.rs [hunk 2]
```diff
@@ -754,7 +740,12 @@ impl<S: std::fmt::Debug + ForkSource> InMemoryNodeInner<S> {
         batch_env: L1BatchEnv,
         system_env: SystemEnv,
         fork_storage: &ForkStorage<S>,
-    ) -> VmExecutionResultAndLogs {
+    ) -> (
+        VmExecutionResultAndLogs,
+        Vec<Call>,
+        ErrorFlags,
+        BootloaderDebug,
+    ) {
```
Comment (dutterbutter) [inline src/node/in_memory.rs:748]: I dont love returning this tuple from here, but unsure how else to make it work

**Comment by ly0va** [comment]
Very cool!

Can we make `data` be displayed in hex instead of Uint8Array? I think it's a bit more concise and readable that way. 

Also, controversial, but not all terminals support emoji. Can we find an ascii/utf8 replacement for :red_circle: ?

**Review by itegulov — COMMENTED** [review]
Sorry for the massive amount of conflicts, completely missed this PR! I think this is still useful and lmk if you need help with rebasing the changes on the new project structure.

**Pull request closed without merging**